tests/test_r3f.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Testing r3f implementation.
+"""
+
+from parlai.nn.r3f import R3FMixin, R3F_NOISE_UNIFORM
+import torch.nn as nn
+import torch
+
+import unittest
+
+GENERATIVE = "generative"
+DECODER_ONLY = "decoder_only"
+# not doing this for encoder only yet, so not making a test case
+
+DUMMY_MODEL_RETURN = (torch.Tensor(0), [])
+
+
+class _DummyModuleWithEmbedding(nn.Module):
+    def __init__(self, expect_noise_this, name, model_type):
+        super(_DummyModuleWithEmbedding, self).__init__()
+        self.expect_noise_this = expect_noise_this
+        self.name = name
+        self.model_type = model_type
+        weight = torch.FloatTensor([[1]])
+        self.norm_embeddings = nn.Embedding.from_pretrained(weight)
+        self.had_first_pass = False
+
+    def forward(self, *args, **kwargs):
+        if self.had_first_pass:
+            val = self.norm_embeddings(torch.LongTensor([0]))
+            if self.expect_noise_this:
+                assert (
+                    val.item() != 1.0
+                ), f"Did not noise embedding of '{self.name}' in '{self.model_type}' when expected"
+            else:
+                assert (
+                    val.item() == 1.0
+                ), f"Noised embedding '{self.name}' in '{self.model_type}' when not expected"
+        self.had_first_pass = True
+
+
+class _DummyGenerativeModel(nn.Module):
+    """
+    Dummy module with an encoder + decoder, and corresponding embeddings.
+    """
+
+    def __init__(self, opt):
+        super(_DummyGenerativeModel, self).__init__()
+        self.opt = opt
+        self.encoder = _DummyModuleWithEmbedding(
+            opt.get('r3f_encoder_noise'), 'encoder', GENERATIVE
+        )
+        self.decoder = _DummyModuleWithEmbedding(
+            opt.get('r3f_decoder_noise'), 'decoder', GENERATIVE
+        )
+
+    def forward(self, *args, **kwargs):
+        self.encoder()
+        self.decoder()
+        return DUMMY_MODEL_RETURN
+
+
+class _DummyDecoderOnlyModelLikeGPT2(nn.Module):
+    """
+    Dummy class to test that R3F code is working properly on modles with only a decoder.
+
+    R3F code finds decoder modules with a regex on the class name (currently looks for
+    "GPT" as in "GPT2")
+    """
+
+    def __init__(self, opt):
+        super(_DummyDecoderOnlyModelLikeGPT2, self).__init__()
+        self.opt = opt
+        self.decoder = _DummyModuleWithEmbedding(
+            opt.get('r3f_decoder_noise'), 'decoder', DECODER_ONLY
+        )
+
+    def forward(self, *args, **kwargs):
+        self.decoder()
+        return DUMMY_MODEL_RETURN
+
+
+######################### Following are wrapper abstractions to work with the fact that R3F is tightly coupled with how models in Parlai are set up.
```

Comment:
:'(

Reply:
Yeah, couldn't figure out how to do it without mocking everything in between.

Reply:
Given that you don't do a test of correctness, we could just do an end to end test? Train for 5 sgd steps...

Reply:
Not sure if we can check for adding noise to encoder/decoder that way.

Actually, I suppose I might be able to string something together via capturing log output, hmm....

Reply:
We do have a capture_stdout and with.assertLogs but I hate using them. Your mock is good.